[2chain] switch to use 2-chain in Genesis module

zekun000 opened the pull request:
Starting from this commit, new genesis would be initialized with two_chain = true that enables 2-chain consensus protocol by default.

<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

Starting from this commit, new genesis would be initialized with two_chain = true that enables 2-chain consensus protocol by default.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

zekun000 commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/1218894192)

JoshLind commented on state-sync/state-sync-v1/src/executor_proxy.rs:
```diff
@@ -638,19 +638,19 @@ mod tests {
             .on_chain_configs
             .get::<OnChainConsensusConfig>()
-            .is_err());
+            .is_ok());
```
`assert(.. is_ok())` 😆

JoshLind commented on language/diem-framework/modules/doc/DiemConsensusConfig.md:
```diff
@@ -15,7 +15,6 @@ DiemConfig, and may be updated by Diem root.
 
 
 <pre><code><b>use</b> <a href="DiemConfig.md#0x1_DiemConfig">0x1::DiemConfig</a>;
-<b>use</b> <a href="DiemTimestamp.md#0x1_DiemTimestamp">0x1::DiemTimestamp</a>;
```
Silly question, but I'm guessing these are auto-generated?

zekun000 replied:
yeah, it's generated by the diem-framework binary

JoshLind commented on language/diem-framework/modules/DiemConsensusConfig.move:
```diff
@@ -27,8 +26,6 @@ module DiemFramework::DiemConsensusConfig {
 
     /// Allows Diem root to update the config.
     public fun set(dr_account: &signer, config: vector<u8>) {
-        DiemTimestamp::assert_operating();
```
Silly question: I'm guessing this is so that we can set the config on genesis itself, and not only after genesis has occurred?

zekun000 replied:
exactly

JoshLind reviewed (approved):
LGTM! 😄

zekun000 commented:
/land

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/1219893174)

zekun000 commented:
/land

dmitri-perelman commented:
@zekun000 silly question: aren't we supposed to see a 1/3 latency decrease in cluster test in this PR? (it's showing the same numbers)

zekun000 commented:
@dmitri-perelman no, under the current perf test, see the discussion back in Feb https://toroeng.slack.com/archives/CH8MKQ60G/p1614019517000900

dmitri-perelman commented:
@zekun000, in [your drawing](https://docs.google.com/drawings/d/1LpQzEeKiR7ZIQtunoVciv2J9Ay47J2zct_v93CK7ecI/edit), the time a transaction waits in the mempool is different for 2-chain and 3-chain commit protocols, which is counter-intuitive. 
Theoretically, the time a transaction waits in the mempool is a function of throughput (hence, the same time for 2-chain and 3-chain), so 2-chain latency would be still better.
That said, I now understand that the latency improvement is hard to see at 100% throughput because latency there is dominated by MP wait time.

Should we have some test of sending 10 transactions and checking their avg latency while the system is idle?

zekun000 commented:
> Theoretically, the time a transaction waits in the mempool is a function of throughput 

You're absolutely right about this, but important part here is that 2-chain *doesn't improve the throughput*, it purely reduces the commit latency. Throughput is determined by how fast the protocol can process `a single block`.

> Should we have some test of sending 10 transactions and checking their avg latency while the system is idle?

We do have the exact test in #changelog, here's the number I got before and after
3-chain: fixed tps 10 : 10 TPS, 756 ms latency, 900 ms p99 latency
2-chain: fixed tps 10 : 10 TPS, 593 ms latency, 700 ms p99 latency

I also gathered latency vs throughput plot here https://toroeng.slack.com/archives/C01P97NFXNY/p1628804260043500

zekun000 commented:
@dmitri-perelman oh your statement is correct that the throughput is the same (I misread it). 

the reason why the wait time is longer for 2-chain vs 3-chain is because 3-chain holds a 3 blocks buffer in consensus (excluded from mempool wait time) while 2-chain holds 2 blocks buffer. so the difference between MP wait time should be the processing time of one block.

in the end, when the system is under load, the latency is only a function of throughput and not related to 2/3 chain commit (e2e latency = MP wait time + consensus latency = how many blocks are ahead/throughput).

dmitri-perelman commented:
Thanks @zekun000! Makes perfect sense, we're on the same page! (The only reason I commented was that in theory 2-chain latency is still "better" than 3-chain latency even under the high load, it's just hard to observe it because the one-block RTT is negligible in comparison with the MP wait time).

I fail to see the "fixed tps 10" in the cluster test output: the [logs](http://kibana.ct-0-k8s-testnet.aws.hlw3truzy4ls.com/app/kibana#/discover?_g=(time:(from:'2021-09-10T19:15:01Z',to:'2021-09-10T19:42:00Z'))) link to kibana finds nothing :). Where do you find this line?

zekun000 commented:
fixed tps experiment is not part of landing blocking suite, it's part of pre-release suite which is reported in #changelog channel https://toroeng.slack.com/archives/CPKBHVA90/p1631337715004500
